designs/SpanMarshallers.md

```diff
@@ -0,0 +1,247 @@
+# Support for Marshalling `(ReadOnly)Span<T>`
+
+As part of the exit criteria for the DllImportGenerator experiment, we have decided to introduce support for marshalling `System.Span<T>` and `System.ReadOnlySpan<T>` into the DllImportGenerator-generated stubs. This document describes design decisions made during the implementation of these marshallers.
+
+## Design 1: "Intrinsic" support for `(ReadOnly)Span<T>`
+
+In this design, the default support for `(ReadOnly)Span<T>` is emitted into the marshalling stub directly and builds on the pattern we enabled for arrays.
+
+### Default behavior
+
+This section describes the behavior of the default emitted-in-stub marshallers.
+
+By default, we will marshal `Span<T>` and `ReadOnlySpan<T>` similarly to array types. When possible, we will pin the `(ReadOnly)Span<T>`'s data and pass that down. When it is not possible to do so, we will try stack allocating scratch space for efficiency or allocate native memory with `Marshal.AllocCoTaskMem` for compatibility with arrays.
+
+To support marshalling from native to managed, we will support the same `MarshalAsAttribute` properties that arrays support today.
+
+When a `(ReadOnly)Span<T>` is marshalled from native to managed, we will allocate a managed array and copy the data from the native memory into the array.
+
+### Empty spans
+
+Since `(ReadOnly)Span<T>` does not have a way to distinguish from an empty collection and a `null` collection, we will provide a dummy pointer for zero-length/`null` spans to match the behavior of zero-length arrays. We will also provide an in-source marshaller that enables marshalling an empty span as `null`.
```

Comment:
I do not think this needs to match behavior of arrays by default, and it makes the behavior unintuitive and the marshaling less efficient than it needs for most users.

I think it would be ok to have this as an optional behavior.

Reply:
This is not just about performance, it is also about intuitive behavior. I think having `default` Span to marshal as null pointer is much more intuitive than having it marshal as a random non-null pointer.

Reply:
And also about consistency between C# `fixed` statement and marshaling behavior.

Reply:
`default(byte[])` will marshal as null pointer today. I expect that there are number of cases where people use null array to pass null pointer into the underlaying API that really needs to see null.

Arrays have a strong distinction between default and zero-length. Span does not have a strong distinction between default and zero-length. We will need to choose which one of those Span cases is going to be less compatible with arrays.

Reply:
So is the expected behavior correct then:

```csharp
var t = new int[0];
Span<int> tt = t;  // Assume should be the same as if 't' were passed.
Span<int> ttt = null; // Assume null will be passed.
Span<int> tttt = new Span<int>(null, 0); // Assume null will be passed.
```

I want to make sure users updating arguments from `T[]` to `Span<T>` don't get into trouble here.

Reply:
The `Span` passed into P/Invoke would be comming from ordinary user code some of the time. The ordinary code working with Spans never makes distinction between `IsEmpty` span with null pointer and `IsEmpty` span with non-null pointer. In other words, the pointer value is irelevant for `IsEmpty` spans. It is why we have choosen to normalize the `IsEmpty` spans into null pointer for pinning. It avoids gotcha with behavior differences for the two different `IsEmpty` spans, and it is also good for security (guaranteed AV when accesing zero-sized buffer).

Reply:
```
var t = new int[0];
Span<int> tt = t;  // Assume should be the same as if 't' were passed.
```
We can make this case work as you have written (same as if `t` were passed) by skipping the normalization and using the raw Span pointer. It would have the risk of behavior differences for the null vs. non-null Spans.

Reply:
It looks like MemoryMarshal.GetReference will give us the behavior described.